packages/dao-dao-macros/src/lib.rs

```diff
@@ -210,6 +210,65 @@ pub fn token_query(metadata: TokenStream, input: TokenStream) -> TokenStream {
     )
 }
 
+/// Adds the necessary fields to an enum such that it implements the
+/// interface needed to be a voting module with a native token.
+///
+/// For example:
+///
+/// ```
+/// use dao_dao_macros::native_token_query;
+/// use cosmwasm_schema::{cw_serde, QueryResponses};
+/// use cosmwasm_std::Addr;
+///
+/// #[native_token_query]
+/// #[cw_serde]
+/// #[derive(QueryResponses)]
+/// enum QueryMsg {}
+/// ```
+///
+/// Will transform the enum to:
+///
+/// ```
+/// enum QueryMsg {
+///     Denom {},
+/// }
+/// ```
+///
+/// Note that other derive macro invocations must occur after this
+/// procedural macro as they may depend on the new fields. For
+/// example, the following will fail becase the `Clone` derivation
+/// occurs before the addition of the field.
+///
+/// ```compile_fail
+/// use dao_dao_macros::native_token_query;
+/// use cosmwasm_schema::{cw_serde, QueryResponses};
+///
+/// #[derive(Clone)]
+/// #[native_token_query]
+/// #[allow(dead_code)]
+/// #[cw_serde]
+/// #[derive(QueryResponses)]
+/// enum Test {
+///     Foo,
+///     Bar(u64),
+///     Baz { foo: u64 },
+/// }
+/// ```
+#[proc_macro_attribute]
+pub fn native_token_query(metadata: TokenStream, input: TokenStream) -> TokenStream {
+    merge_variants(
+        metadata,
+        input,
+        quote! {
+        enum Right {
+            #[returns(::dao_interface::voting::DenomResponse)]
+            Denom {}
```

Comment:
This is great. `dao-voting-token-staked` DAOs do have a "Token Contract", `cw-tokenfactory-issuer` which is used for interacting with the Token Factory token and provides some extra features.

Now it's possible to get that from the denom, but requires having to parse it.

Don't feel too strongly here, just wanted to mention it. Maybe `cw20_token_query` could be just `token_contract` query? Then you could have both?

Reply:
i think one of us is misunderstanding the other 😅

the `token_contract` in `dao-voting-token-staked` and `dao-voting-cw20-staked` are fundamentally different. the `voting_module_token` variant is used exclusively when instantiating with a proposal deposit, and the pre-propose module setting up the proposal deposit needs to query to find the denom that the voting module uses (either native denom or cw20 contract) to fit into the `CheckedDenom` enum and save it in the config. this has nothing to do with the token capabilities, it just needs to retrieve the actual token being used during instantiation so that it can save the right denom for future proposal deposit validations.

Reply:
Never mind me, I see the query is still there in the `msg.rs` file.